File: lib/profile-metadata-policy.ts
import { getMatrixClient } from './matrix';
import { checkedConversationState, canEditConversationState } from './channel-administration';
import type { Profile } from './community';

export const profileMetadataPolicyEvent = 'io.tavern.server.profile_policy';
export const profileBioLimits = [0, 160, 500, 1000] as const;
export const profileStatusLimits = [0, 40, 80, 160] as const;
export type ProfileMetadataPolicy = { version: 1; enabled: boolean; allowLinks: boolean; allowCustomFields: boolean; maxBioLength: number; maxStatusLength: number };
export type ApplicableProfileMetadataPolicy = ProfileMetadataPolicy & { status: 'ready' | 'unavailable'; serverIds: string[]; reason?: string };
export const emptyProfileMetadataPolicy = (): ProfileMetadataPolicy => ({ version: 1, enabled: false, allowLinks: true, allowCustomFields: true, maxBioLength: 1000, maxStatusLength: 160 });
const privateType = 'io.tavern.private_thread';
const object = (value: unknown): value is Record<string, any> => !!value && typeof value === 'object' && !Array.isArray(value);
const id = (value: unknown, prefix: string): value is string => typeof value === 'string' && value.startsWith(prefix) && value.length >= 2 && value.length <= 1024 && !/[\s\x00-\x1f\x7f]/.test(value);
const content = (room: any, type: string, key = '') => room?.currentState.getStateEvents(type, key)?.getContent() || {};
const event = (room: any, type: string) => room?.currentState.getStateEvents(type, '');
const unavailable = 'The current server profile rules are unavailable. Wait for the server to sync or ask its owner to repair them.';

export function parseProfileMetadataPolicy(value: unknown): ProfileMetadataPolicy | null {
  if (!object(value) || value.version !== 1 || ['enabled', 'allowLinks', 'allowCustomFields'].some(key => typeof value[key] !== 'boolean')
    || !(profileBioLimits as readonly unknown[]).includes(value.maxBioLength) || !(profileStatusLimits as readonly unknown[]).includes(value.maxStatusLength)
    || Object.keys(value).some(key => !['version', 'enabled', 'allowLinks', 'allowCustomFields', 'maxBioLength', 'maxStatusLength', 'io.tavern.previous_event'].includes(key))
    || value['io.tavern.previous_event'] !== undefined && value['io.tavern.previous_event'] !== null && !id(value['io.tavern.previous_event'], '$')) return null;
  return { version: 1, enabled: value.enabled, allowLinks: value.allowLinks, allowCustomFields: value.allowCustomFields, maxBioLength: value.maxBioLength, maxStatusLength: value.maxStatusLength };
}
const parseSavedPolicy = (value: unknown) => object(value) && Object.hasOwn(value, 'io.tavern.previous_event') ? parseProfileMetadataPolicy(value) : null;
export function readServerProfileMetadataPolicy(serverId: string) {
  const saved = event(getMatrixClient()?.getRoom(serverId), profileMetadataPolicyEvent);
  return saved ? parseSavedPolicy(saved.getContent()) : emptyProfileMetadataPolicy();
}
function parentIds(room: any): string[] {
  const parents = room.currentState.getStateEvents('m.space.parent').filter((item: any) => item.getContent().canonical === true && item.getContent().via?.length);
  if (parents.length > 32 || parents.some((item: any) => !id(item.getStateKey(), '!'))) throw new Error(unavailable);
  return [...new Set<string>(parents.map((item: any) => item.getStateKey()))];
}
function sourceId(room: any): string | null {
  const creation = content(room, 'm.room.create');
  if (creation.type !== privateType) return null;
  const binding = creation[privateType];
  if (creation['m.federate'] !== false || !object(binding) || binding.version !== 1 || Object.keys(binding).sort().join(',') !== 'source_event_id,source_room_id,version'
    || !id(binding.source_room_id, '!') || binding.source_event_id !== '' && !id(binding.source_event_id, '$')) throw new Error(unavailable);
  return binding.source_room_id;
}
function scopedRooms(roomId: string, lookup: (id: string) => any) {
  const original = lookup(roomId); if (!original) throw new Error(unavailable);
  const source = sourceId(original), room = source ? lookup(source) : original;
  if (!room || !event(room, 'm.room.create') || source && ['m.space', privateType].includes(content(room, 'm.room.create').type)) throw new Error(unavailable);
  const parents = parentIds(room).map(parentId => {
    const parent = lookup(parentId); if (!parent || !event(parent, 'm.room.create')) throw new Error(unavailable);
    return content(parent, 'm.space.child', room.roomId).via?.length ? parent : null;
  }).filter(Boolean);
  return { original, source: room, rooms: [room, ...parents] };
}
function combined(rooms: any[]): ApplicableProfileMetadataPolicy {
  const result: ApplicableProfileMetadataPolicy = { ...emptyProfileMetadataPolicy(), status: 'ready', serverIds: [] };
  for (const room of rooms) {
    const saved = event(room, profileMetadataPolicyEvent); if (!saved) continue;
    const policy = parseSavedPolicy(saved.getContent()), create = content(room, 'm.room.create');
    if (!policy || policy.enabled && (create.type !== 'm.space' || create['m.federate'] !== false)) throw new Error(unavailable);
    if (!policy.enabled) continue;
    result.enabled = true; result.serverIds.push(room.roomId);
    result.allowLinks &&= policy.allowLinks; result.allowCustomFields &&= policy.allowCustomFields;
    result.maxBioLength = Math.min(result.maxBioLength, policy.maxBioLength); result.maxStatusLength = Math.min(result.maxStatusLength, policy.maxStatusLength);
  }
  return result;
}
export function profilePolicyForRoom(roomId: string): ApplicableProfileMetadataPolicy {
  try { const client = getMatrixClient(); return combined(scopedRooms(roomId, value => client?.getRoom(value)).rooms); }
  catch { return { ...emptyProfileMetadataPolicy(), status: 'unavailable', serverIds: [], reason: unavailable }; }
}
export function profileMetadataPolicyError(profile: Pick<Profile, 'bio' | 'status' | 'links' | 'fields'> & Partial<Pick<Profile, 'statusEmoji'>>, rules: ApplicableProfileMetadataPolicy): string | null {
  if (rules.status !== 'ready') return rules.reason || unavailable;
  if (!rules.enabled) return null;
  const reasons: string[] = [];
  if (!rules.allowLinks && profile.links.length) reasons.push('Profile links are disabled');
  if (!rules.allowCustomFields && profile.fields.length) reasons.push('Custom profile fields are disabled');
  if (profile.bio.length > rules.maxBioLength) reasons.push(`Bio must be at most ${rules.maxBioLength} characters`);
  if (!rules.maxStatusLength && (profile.status.length || profile.statusEmoji)) reasons.push('Status and status emoji are disabled');
  else if (profile.status.length > rules.maxStatusLength) reasons.push(`Status must be at most ${rules.maxStatusLength} characters`);
  return reasons.length ? reasons.join('. ') + '. Your draft has been kept.' : null;
}
function boundedText(value: string, length: number) { const text = value.slice(0, length); return /[\ud800-\udbff]$/.test(text) ? text.slice(0, -1) : text; }
export function applyProfileMetadataPolicy<T extends Profile>(profile: T, rules: ApplicableProfileMetadataPolicy): T {
  if (rules.status !== 'ready') throw new Error(rules.reason || unavailable);
  return { ...profile, links: rules.enabled && !rules.allowLinks ? [] : profile.links.map(link => ({ ...link })), fields: rules.enabled && !rules.allowCustomFields ? [] : profile.fields.map(field => ({ ...field })),
    bio: boundedText(profile.bio, rules.enabled ? rules.maxBioLength : 1000), status: boundedText(profile.status, rules.enabled ? rules.maxStatusLength : 160),
    statusEmoji: rules.enabled && !rules.maxStatusLength ? '' : profile.statusEmoji, statusUntil: rules.enabled && !rules.maxStatusLength ? 0 : profile.statusUntil };
}
/** Presentation only: older membership events remain readable through Matrix. */
export function visibleProfileMetadata<T extends Profile>(profile: T, roomId: string): T {
  const rules = profilePolicyForRoom(roomId);
  return rules.status === 'ready' ? applyProfileMetadataPolicy(profile, rules) : { ...profile, bio: '', status: '', statusEmoji: '', statusUntil: 0, links: [], fields: [] };
}
function roomFromState(roomId: string, values: any[]) {
  if (!Array.isArray(values) || values.length > 50000) throw new Error(unavailable);
  const entries = new Map<string, any>();
  for (const value of values) {
    if (!object(value) || typeof value.type !== 'string' || typeof value.state_key !== 'string' || !object(value.content)) throw new Error(unavailable);
    const key = value.type + '\0' + value.state_key; if (entries.has(key)) throw new Error(unavailable);
    entries.set(key, { getContent: () => value.content, getStateKey: () => value.state_key, getId: () => value.event_id, getSender: () => value.sender, type: value.type });
  }
  return { roomId, currentState: { getStateEvents: (type: string, key?: string) => key === undefined ? [...entries.values()].filter(item => item.type === type) : entries.get(type + '\0' + key) || null } };
}
export async function checkProfileMetadataPublication(roomId: string, profile: Profile, expectedClient = getMatrixClient(), project = false) {
  const client = expectedClient, actor = client?.getUserId();
  const current = () => client && getMatrixClient() === client && client.getUserId() === actor && client.getRoom(roomId)?.getMyMembership() === 'join';
  if (!client || !actor || !current()) throw new Error('Your account or conversation membership changed.');
  const rooms = new Map<string, ReturnType<typeof roomFromState>>();
  const load = async (identity: string) => { if (!current()) throw new Error('Your account or conversation membership changed.'); const room = roomFromState(identity, await client.roomState(identity)); if (!current()) throw new Error('Your account or conversation membership changed.'); rooms.set(identity, room); return room; };
  const original = await load(roomId), source = sourceId(original), base = source ? await load(source) : original;
  const parents = parentIds(base); // Bound the entire parent set before any parent request.
  for (const parent of parents) if (!rooms.has(parent)) await load(parent);
  const scope = scopedRooms(roomId, identity => rooms.get(identity));
  if ([scope.original, ...scope.rooms].some(room => content(room, 'm.room.member', actor).membership !== 'join')) throw new Error('Join the conversation and its current source servers before publishing a profile.');
  const rules = combined(scope.rooms), publication = project ? applyProfileMetadataPolicy(profile, rules) : profile, failure = profileMetadataPolicyError(publication, rules);
  if (failure) throw new Error(failure);
  if (!current()) throw new Error('Your account or conversation membership changed.');
  return { client, actor, membership: content(original, 'm.room.member', actor), rules, profile: publication };
}
export function canManageProfileMetadataPolicy(serverId: string) {
  const client = getMatrixClient(), room = client?.getRoom(serverId), creation = event(room, 'm.room.create');
  if (!event(room, 'io.tavern.roles') && creation?.getSender() !== client?.getUserId()) return false;
  return !!(room?.isSpaceRoom() && creation?.getContent()['m.federate'] === false && canEditConversationState(serverId, profileMetadataPolicyEvent));
}
export async function saveProfileMetadataPolicy(serverId: string, value: ProfileMetadataPolicy, previous: ProfileMetadataPolicy | null) {
  const next = parseProfileMetadataPolicy(value), owner = getMatrixClient();
  if (!next || !canManageProfileMetadataPolicy(serverId)) throw new Error('Choose valid profile rules and check your server permissions.');
  const client = await checkedConversationState(serverId, profileMetadataPolicyEvent);
  if (client !== owner) throw new Error('Your account changed. Reopen these settings.');
  const fresh = roomFromState(serverId, await client.roomState(serverId)), saved = event(fresh, profileMetadataPolicyEvent);
  const current = saved ? parseSavedPolicy(saved.getContent()) : emptyProfileMetadataPolicy();
  if (JSON.stringify(current) !== JSON.stringify(previous)) throw new Error('Profile rules changed. Your draft is preserved. Reload the current rules before saving.');
  if (saved && !id(saved.getId(), '$')) throw new Error('The current profile rule revision is unavailable.');
  const checked = await checkedConversationState(serverId, profileMetadataPolicyEvent);
  if (checked !== owner || getMatrixClient() !== owner || !canManageProfileMetadataPolicy(serverId)) throw new Error('Your account or server permissions changed.');
  await client.sendStateEvent(serverId, profileMetadataPolicyEvent as any, { ...next, 'io.tavern.previous_event': saved?.getId() ?? null }, '');
  if (getMatrixClient() !== owner) throw new Error('Your account changed while the profile rules were saving.');
  return next;
}

File: synapse_modules/profile_metadata_policy.py
"""Opt-in plaintext profile-extension bounds; never classify encrypted content."""
from collections.abc import Mapping
import json
import re
from urllib.parse import urlsplit

PROFILE_POLICY = 'io.tavern.server.profile_policy'
PROFILE = 'io.tavern.profile'
PRIVATE = 'io.tavern.private_thread'
BIO_LIMITS = frozenset({0, 160, 500, 1000})
STATUS_LIMITS = frozenset({0, 40, 80, 160})
MAX_BYTES = 32768
TEXT_LIMITS = {'name': 60, 'bio': 1000, 'pronouns': 50, 'timezone': 80, 'language': 30,
               'status': 160, 'statusEmoji': 16}


class ProfilePolicyDenied(Exception):
    pass


def value(state, kind, key=''):
    event = state.get((kind, key))
    return event.content if event else {}


def snapshot(data):
    # Synapse1.160 event.content is a Rust-backed Mapping, not a Python dict.
    if isinstance(data, Mapping):
        return {key: snapshot(item) for key, item in data.items()}
    if isinstance(data, (list, tuple)):
        return [snapshot(item) for item in data]
    if data is None or type(data) in (str, bool, int, float):
        return data
    raise ValueError('Invalid JSON value')


def valid_settings(data):
    return (isinstance(data, Mapping)
            and set(data) == {'version', 'enabled', 'allowLinks', 'allowCustomFields', 'maxBioLength', 'maxStatusLength', 'io.tavern.previous_event'}
            and type(data.get('version')) is int and data['version'] == 1
            and all(type(data.get(key)) is bool for key in ('enabled', 'allowLinks', 'allowCustomFields'))
            and type(data.get('maxBioLength')) is int and data['maxBioLength'] in BIO_LIMITS
            and type(data.get('maxStatusLength')) is int and data['maxStatusLength'] in STATUS_LIMITS
            and (data['io.tavern.previous_event'] is None or isinstance(data['io.tavern.previous_event'], str)
                 and bool(re.fullmatch(r'\$[^\s\x00-\x1f\x7f]{1,1023}', data['io.tavern.previous_event']))))


def text(value, maximum, multiline=True):
    return (isinstance(value, str) and len(value.encode('utf-16-le', errors='surrogatepass')) // 2 <= maximum
            and not re.search(r'[\x00-\x08\x0b\x0c\x0e-\x1f\x7f\ud800-\udfff]' if multiline else r'[\x00-\x20\x7f\ud800-\udfff]', value))


def safe_url(value, mxc=False):
    if not text(value, 1024 if mxc else 1000, multiline=False) or not value or '\\' in value:
        return False
    try:
        parsed = urlsplit(value)
        if (parsed.scheme not in ({'mxc'} if mxc else {'http', 'https'}) or not parsed.hostname
                or parsed.username is not None or parsed.password is not None):
            return False
        if parsed.port is not None and not 1 <= parsed.port <= 65535:
            return False
        return not mxc or (not parsed.query and not parsed.fragment and bool(re.fullmatch(r'/[^/]+', parsed.path)))
    except (ValueError, UnicodeError):
        return False


def profile_error(profile, room_id, is_space):
    if (not isinstance(profile, Mapping) or type(profile.get('version')) is not int or profile['version'] != 1
            or set(profile) - ({'version', 'avatar', 'banner', 'accent', 'statusUntil', 'links', 'fields', 'serverOverride'} | set(TEXT_LIMITS))):
        return 'Use a supported version1 profile extension without extra fields.'
    for key, maximum in TEXT_LIMITS.items():
        if not text(profile.get(key, ''), maximum):
            return 'The profile ' + key + ' contains unsupported text or exceeds its length limit.'
    for key in ('avatar', 'banner'):
        item = profile.get(key, '')
        if item != '' and not safe_url(item, mxc=True):
            return 'Profile artwork must use a bounded Matrix media URI.'
    accent = profile.get('accent', '')
    if not isinstance(accent, str) or accent and not re.fullmatch(r'#[0-9A-Fa-f]{6}', accent):
        return 'Choose a valid profile accent color.'
    until = profile.get('statusUntil', 0)
    if type(until) is not int or not 0 <= until <= 9007199254740991:
        return 'Choose a valid profile status expiration.'
    override = profile.get('serverOverride', '')
    if not isinstance(override, str) or override and (not is_space or override != room_id):
        return 'A server profile override must belong to this server.'
    links, fields = profile.get('links', []), profile.get('fields', [])
    if not isinstance(links, (list, tuple)) or len(links) > 5 or not isinstance(fields, (list, tuple)) or len(fields) > 8:
        return 'Keep at most five profile links and eight custom fields.'
    for link in links:
        if not isinstance(link, Mapping) or set(link) != {'label', 'url'} or not text(link['label'], 60) or not safe_url(link['url']):
            return 'Profile links require bounded HTTP or HTTPS URLs without credentials.'
    for field in fields:
        if (not isinstance(field, Mapping) or set(field) != {'label', 'value'}
                or not text(field['label'], 60) or not field['label'].strip()
                or not text(field['value'], 300) or not field['value'].strip()):
            return 'Custom profile fields require bounded, nonempty labels and values.'
    if len(json.dumps(snapshot(profile), ensure_ascii=False, separators=(',', ':')).encode()) > MAX_BYTES:
        return 'The profile extension exceeds 32 KiB.'
    return None


class ProfileMetadataPolicy:
    def __init__(self, api, model):
        self.api, self.model = api, model

    @staticmethod
    def owner(state, actor):
        create = state.get(('m.room.create', ''))
        return bool(create and create.sender == actor and create.content.get('type') == 'm.space')

    async def scopes(self, room_id, state):
        creation = value(state, 'm.room.create')
        if creation.get('type') == PRIVATE:
            try:
                from private_thread import binding
            except ImportError:
                from synapse_modules.private_thread import binding
            data = binding(creation)
            if not data:
                raise ValueError('Invalid private source')
            room_id = data['source_room_id']
            state = await self.api.get_room_state(room_id)
            source = value(state, 'm.room.create')
            if source.get('m.federate') is not False or source.get('type') in (PRIVATE, 'm.space'):
                raise ValueError('Invalid private source')
        found = [(room_id, state)] if (PROFILE_POLICY, '') in state else []
        parents = {key for (kind, key), item in state.items() if kind == 'm.space.parent' and item.content.get('canonical') is True and item.content.get('via')}
        if len(parents) > 32:
            raise ValueError('Too many server scopes')
        for parent_id in sorted(parents):
            parent = await self.api.get_room_state(parent_id)
            if (PROFILE_POLICY, '') in parent and value(parent, 'm.space.child', room_id).get('via'):
                found.append((parent_id, parent))
        return found

    def may_configure(self, event, state):
        creation, previous = value(state, 'm.room.create'), state.get((PROFILE_POLICY, ''))
        if (getattr(event, 'state_key', None) != '' or not valid_settings(event.content)
                or creation.get('type') != 'm.space' or creation.get('m.federate') is not False
                or event.content['io.tavern.previous_event'] != (previous.event_id if previous else None)
                or value(state, 'm.room.member', event.sender).get('membership') != 'join'):
            return False
        powers = value(state, 'm.room.power_levels')
        if not isinstance(powers.get('events', {}), Mapping):
            return False
        threshold = powers.get('events', {}).get(PROFILE_POLICY, powers.get('state_default', 50))
        power = self.model.native_member_power(state, event.sender)
        if type(threshold) is not int or power is None or power < threshold:
            return False
        policy = value(state, self.model.POLICY)
        if (self.model.POLICY, '') in state:
            return self.model.valid_policy(policy) and 'manage_server' in self.model.permissions(policy, event.sender)
        return self.owner(state, event.sender)

    @staticmethod
    def fingerprint(scopes):
        return {identity: (current[(PROFILE_POLICY, '')].event_id, snapshot(value(current, PROFILE_POLICY)),
                           getattr(current.get(('m.room.create', '')), 'sender', None), snapshot(value(current, 'm.room.create')))
                for identity, current in scopes}

    def validate_publication(self, event, state, active):
        if event.sender != event.state_key:
            raise ProfilePolicyDenied('Only the account owner can publish its extended membership profile.')
        profile = event.content[PROFILE]
        error = profile_error(profile, event.room_id, value(state, 'm.room.create').get('type') == 'm.space')
        if error:
            raise ProfilePolicyDenied(error)
        if profile.get('links') and any(not policy['allowLinks'] for policy in active):
            raise ProfilePolicyDenied('This server does not allow profile links. Remove the link list before publishing here.')
        if profile.get('fields') and any(not policy['allowCustomFields'] for policy in active):
            raise ProfilePolicyDenied('This server does not allow custom profile fields. Remove them before publishing here.')
        for field, setting in (('bio', 'maxBioLength'), ('status', 'maxStatusLength')):
            maximum = min(policy[setting] for policy in active)
            if not text(profile.get(field, ''), maximum):
                raise ProfilePolicyDenied('This server limits profile ' + field + ' to ' + str(maximum) + ' characters.')
            if field == 'status' and maximum == 0 and profile.get('statusEmoji'):
                raise ProfilePolicyDenied('This server disables profile status. Remove the status emoji before publishing here.')

    async def publication(self, event, state):
        scopes = await self.scopes(event.room_id, state)
        before = self.fingerprint(scopes)
        active = []
        for _, current in scopes:
            config = value(current, PROFILE_POLICY)
            if not valid_settings(config):
                raise ProfilePolicyDenied('The server profile policy needs repair. Remove the extended profile or contact its owner.')
            if not config['enabled']:
                continue
            creation = value(current, 'm.room.create')
            if creation.get('type') != 'm.space' or creation.get('m.federate') is not False:
                raise ProfilePolicyDenied('The server profile policy has an invalid scope.')
            active.append(config)
        if active:
            self.validate_publication(event, state, active)
        # Refresh even if the observed state was off/missing: an owner can
        # enable a requirement or attach a new canonical parent during a read.
        current = await self.api.get_room_state(event.room_id)
        if value(state, 'm.room.create').get('type') == PRIVATE and snapshot(value(current, 'm.room.create')) != snapshot(value(state, 'm.room.create')):
            raise ProfilePolicyDenied('The private discussion source changed. Reopen the discussion.')
        if before != self.fingerprint(await self.scopes(event.room_id, current)):
            raise ProfilePolicyDenied('The server profile requirements changed. Reload the profile before publishing.')
        return True

    async def check(self, event, state):
        kind, key = event.type, getattr(event, 'state_key', None)
        if kind == 'm.room.member' and event.content.get('membership') == 'leave':
            return True  # Neither self-leave nor moderation becomes a profile publication.
        try:
            if kind == PROFILE_POLICY:
                if not self.may_configure(event, state):
                    return False
                return self.may_configure(event, await self.api.get_room_state(event.room_id))
            if kind == 'm.room.redaction':
                target = getattr(event, 'redacts', None) or event.content.get('redacts')
                original = await self.api._store.get_event(target, allow_none=True) if target else None
                if original and original.type == PROFILE_POLICY:
                    return False
                if original and original.type in ('m.space.parent', 'm.space.child'):
                    parent = state if original.type == 'm.space.child' else await self.api.get_room_state(original.state_key)
                    if (PROFILE_POLICY, '') in parent:
                        return False
            if kind in ('m.space.parent', 'm.space.child') and value(state, kind, key).get('via'):
                previous = value(state, kind, key)
                removing = not event.content.get('via') or kind == 'm.space.parent' and previous.get('canonical') is True and event.content.get('canonical') is not True
                if removing:
                    parent = state if kind == 'm.space.child' else await self.api.get_room_state(key)
                    if (PROFILE_POLICY, '') in parent and not self.owner(parent, event.sender):
                        return False
            if kind == 'm.room.member' and event.content.get('membership') == 'join' and PROFILE in event.content:
                return await self.publication(event, state)
            return True
        except ProfilePolicyDenied:
            raise
        except Exception:
            raise ProfilePolicyDenied('Server profile requirements are temporarily unavailable. Retry or remove the extended profile.') from None
